README
HTML-Field version 1.19
=======================

The HTML::Field set of modules creates objects that represent HTML form fields
which try to make it easier to interact with CGI objects, databases, and 
HTML::Template objects. 

The objective of an HTML::Field object is to know how to write its own HTML, 
how to get its value out of a CGI object or from a hash,
how to add their value to a hash suitable for passing into a HTML::Template
or into a SQL::Abstract object, for example, and thus re-use some of the code
which is typically repeated several times in a CGI script.

This bundle includes also HTML::FieldForm, which is a very simple module to 
manage sets of HTML::Field objects.

INSTALLATION

To install these modules type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This set of modules depends only on standard modules 
distributed with Perl. However, to run the complete test suite you need
to have HTML::TokeParser and Test::Exception. Most test files are skipped
unless one or both modules are installed.

COPYRIGHT AND LICENCE

Copyright (C) 2009 by Julio Fraire, julio.fraire@gmail.com

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.3 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension HTML::Field.

# $Id: Field.pm,v 1.19 2009/11/26 00:37:41 julio Exp $
# $Log: Field.pm,v $
# Revision 1.19  2009/11/26 00:37:41  julio
# Corrected test requirement in Textfield.t
#
# Revision 1.18  2009/06/27 22:46:03  julio
# Removed attribute 'wrap' from Textarea
#
# Revision 1.17  2009/05/27 23:21:45  julio
# Updated $VERSION
#
# Revision 1.16  2009/05/27 22:51:10  julio
# If an object other than CGI is given to the value method, it is
# assumed to work as a hash reference. This includes hash references.
# This is to work with DBIx::DataModel based classes.
# Miscellaneous changes to the documentation.
#
# Revision 1.15  2009/05/01 22:05:31  julio
# Fixed problem with hidden fields' readonly_html.
#
# Revision 1.14  2009/05/01 21:50:23  julio
# Fixed bug in checkbox. It was not returning the html field correctly
# when the field was valued.
#
# Revision 1.13  2009/05/01 05:13:29  julio
# Completed documentation on each subclass and fixed Hidden, Textfield
# and Checkbox to conform to the new docs.
#
# Revision 1.12  2009/04/28 20:18:45  julio
# Removed Radiobutton.
#
# Revision 1.11  2009/04/28 20:15:54  julio
# Added HTML::Field::Radio
#
# Revision 1.10  2009/04/27 04:13:31  julio
# After a looong time!
#
# Revision 1.9  2007/10/13 03:51:30  julio
# HTML::Field::Hidden no longer needs to know its value at creation time.
#
# Revision 1.8  2007/09/27 05:59:42  julio
# Added documentation regarding 'default' for HTML::Field::Checkbox.
# Fixed redundant condition in _editable_html.
#
# Revision 1.7  2007/09/26 04:41:22  julio
# Added xhtml switch! I still need to do tests
# and documentation; the current tests passed OK.
#
# Revision 1.6  2007/08/12 02:24:58  julio
# Added method creation_html, docs and tests
#
# Revision 1.5  2007/08/10 02:57:17  julio
# Integrated POD into this file
#
# Revision 1.4  2007/08/10 01:49:45  julio
# Only ran perltidy on this file
#
# Revision 1.3  2007/08/10 01:26:32  julio
# Added logic to implement primary_key and auto into text fields
#
# Revision 1.2  2006/12/10 01:33:11  julio
# Improved test coverage for parent class
#
# Revision 1.1  2006/12/10 00:34:52  julio
# Initial revision
#
# Revision 0.0  2006/12/01 03:26:05  julio
# Initial revision

0.01  Tue Mar 21 23:09:02 2006
	- original version; created by h2xs 1.23 with options
		-XA -n HTML::Field



lib/HTML/FieldForm.pm
package HTML::FieldForm;

use HTML::Field;
use Carp;
use strict;

# Constructor method. Shall receive an array ref of HTML::Field objects.
sub new {
    my $class   = shift;
    my $fields  = {@_};
    
    my %self;
    foreach my $name (keys %$fields) {
        my $type = $fields->{$name}[0];
        my $args = $fields->{$name}[1];
        $args->{name} = $name;
        my $field;
        eval {
            $field = HTML::Field->new($type, %$args);
        };
        croak "Error while creating field $name (class $type): $@"
            if $@;
        $self{$name} = $field;
    }
    return bless \%self, $class;
}

# This method will return the names of all the fields in the form:
sub names {
    my $self = shift;
    return keys %$self;
}

# This method shall receive a hash reference or CGI object to be passed to
# the different Field objects in order for them to read their values.
# Returns nothing.
sub set_values {
    my $self   = shift;
    my $values = shift;
    foreach my $field ( keys %$self ) {
        $self->{$field}->value($values);
    }
    return;
}

# This method shall return the values of the different Field objects as a 
# hash. It is planned to be useful for generating SQL statements by
# SQL::Abstract.
sub get_values {
    my $self   = shift;
    my %result = ();
    foreach my $field ( keys %$self ) {
        $result{ $field } = $self->{$field}->value;
    }
    return %result;
}

# This method will take every field value to its default. It returns 
# nothing.
sub reset_values {
    my $self = shift;
    foreach my $field ( keys %$self ) {
        $self->{$field}->reset_value;
    }
    return;
}

# This method will return a hash containing the editable html 
# (form elements).
# The keys are the names normally given to the HTML::Template parameters or
# form elements. Note that a single field might use several form elements
# (a date or a radio buttons group, for example)
sub editable_html {
    my $self   = shift;
    my %result = ();
    foreach my $field ( keys %$self ) {
        $result{ $field } = $self->{$field}->editable_html;
    }
    return %result;
}

# This method will return a hash containing the read-only html (form elements).
# The keys are the names normally given to the HTML::Template parameters or
# form elements. Note that a single field might use several form elements
# (a date or a radio buttons group, for example)
sub readonly_html {
    my $self   = shift;
    my %result = ();
    foreach my $field ( keys %$self ) {
        $result{ $field } = $self->{$field}->readonly_html;
    }
    return %result;
}

# This method will return a hash containing the editable html (form elements).
# The keys are the names normally given to the HTML::Template parameters or
# form elements. Note that a single field might use several form elements
# (a date or a radio buttons group, for example).
sub creation_html {
    my $self = shift;
    my %result = ();
    foreach my $field ( keys %$self ) {
        $result{ $field } = $self->{$field}->creation_html;
    }
    return %result;
}

# Sets xhtml output or plain html output from HTML::Field. Default is HTML.
# Returns $self to enable chained calls
sub set_xhtml {
    my ($self, $xhtml) = @_;    
    HTML::Field->xhtml($xhtml);
    return $self;
}

# If used, it will add id="name" on every field. Useful for CSS styling
# and using <label>.
# Returns $self to enable chained calls
sub add_id {
    my ($self, $ids) = @_;
    
    foreach my $field (keys %$self) {
        $self->{ $field }->id( $self->{$field}->name );
    }
    
    return $self;
}

1;

=pod

=head1 NAME

FieldForm -- Handle HTML forms with Field objects

=head1 SYNOPSIS

   use HTML::FieldForm;
   use HTML::Field;
   
   # Object constructor
   my $form = HTML::FieldForm->new(    
     part_number      => [ 'Textfield', {
                            primary_key => 1,
                       } ],
     description      => [ 'Textarea', {
                            cols        => 60,
                            rows        => 6,
                       } ],
     purchasing_code =>  [ 'Select', {
                            options     => \@purchasing_codes,
                            labels      => \%purch_codes_labels,
                       } ],
   );
   
   # The returned object is a hash of HTML::Field objects keyed by name:
   my $part_number = $form->{part_number};
   
   # Set the value of the fields from a hash ref
   my %values = (part_number => '16-0021', description => 'Brushplate');
   $form->set_values(\%values);

   # Set the value of the fields from a CGI object
   my $query = new CGI;
   $form->set_values($query);

   # Get the values of all the fields in a hash
   my %data = $form->get_values;
 
   # Reset all the field values to their defaults
   $form->reset_values;
   
   # Get a hash with editable or read only html (fieldname => html)
   # Specially useful for HTML::Template
   my %editable = $form->editable_html;
   my %readonly = $form->readonly_html;
   my %creation = $form->creation_html;
   
   # Request XHTML instead of HTML
   $field_form->set_xhtml(1); # Gets xhtml
   $field_form->set_xhtml(0); # Gets html (default)
   
   # Request the addition of an ID selector for each form field 
   # (same as its name)
   $field_form->add_id;
   
   # These two last methods return the HTML::FieldForm object, so that
   # you can say:
   %editable = $field_form->xhtml(1)->add_id->editable_html;
   
=head1 DESCRIPTION

The goal of this simple module is to perform common actions over a set
of HTML form fields in order to get rid of some of the tedious of scripting
CGI.

=head1 AUTHOR

Julio Fraire, E<lt>jfraire@gmail.com<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 by Julio Fraire

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.3 or,
at your option, any later version of Perl 5 you may have available.

=cut

# $Id: FieldForm.pm,v 0.6 2009/06/05 05:27:50 julio Exp $
# $Log: FieldForm.pm,v $
# Revision 0.6  2009/06/05 05:27:50  julio
# Added set_xhtml and add_id methods.
#
# Revision 0.5  2009/06/05 00:42:46  julio
# Added use HTML::Field, since the new "new" method will create
# the HTML::Field objects itself and there would be no need to use
# HTML::Field from user code.
#


